opentelemetry-sdk/src/runtime.rs

```diff
@@ -6,42 +6,57 @@
 //! [Tokio]: https://crates.io/crates/tokio
 //! [async-std]: https://crates.io/crates/async-std
 
-use futures_util::{future::BoxFuture, stream::Stream};
+use futures_util::stream::{unfold, Stream};
 use std::{fmt::Debug, future::Future, time::Duration};
 use thiserror::Error;
 
 /// A runtime is an abstraction of an async runtime like [Tokio] or [async-std]. It allows
-/// OpenTelemetry to work with any current and hopefully future runtime implementation.
+/// OpenTelemetry to work with any current and hopefully future runtime implementations.
 ///
 /// [Tokio]: https://crates.io/crates/tokio
 /// [async-std]: https://crates.io/crates/async-std
+///
+/// # Note
+///
+/// OpenTelemetry expects a *multithreaded* runtime because its types can move across threads.
+/// For this reason, this trait requires the `Send` and `Sync` bounds. Single-threaded runtimes
+/// can implement this trait in a way that spawns the tasks on the same thread as the calling code.
 #[cfg(feature = "experimental_async_runtime")]
 pub trait Runtime: Clone + Send + Sync + 'static {
-    /// A future stream, which returns items in a previously specified interval. The item type is
-    /// not important.
-    type Interval: Stream + Send;
-
-    /// A future, which resolves after a previously specified amount of time. The output type is
-    /// not important.
-    type Delay: Future + Send + Unpin;
-
-    /// Create a [futures_util::stream::Stream], which returns a new item every
-    /// [std::time::Duration].
-    fn interval(&self, duration: Duration) -> Self::Interval;
```

Comment:
Does the removal of `interval` now restrict the runtimes to bring their own optimized constructs. Eg, we were able to use the `tokio_stream::wrappers::IntervalStream` for tokio runtime. But now, all runtimes now must use the generic unfold+delay implementation provided by `to_interval_stream`? Do we need any performance comparison between the original runtime-specific interval implementations and the new unfold+delay approach ?

Reply:
My thinking here is that for the use-cases of OTel, the interval is relatively high, in case of metrics 60 seconds. While individual runtimes can provide their own optimized stream, it's unlikely this will bring any measurable perf benefit. 

the interval can really be replaced with the following to completely avoid streams:

```rust
loop {
   runtime.delay(...).await;
}
```

I can try to provide some benchmarks, but this is async code which is kinda difficult to measure correctly.

Reply:
> for the use-cases of OTel, the interval is relatively high, in case of metrics 60 seconds

Logs export frequency is 1 seconds.

(Not saying it is an issue, just checking if that makes any difference)